arch: Added initial OpenRISC port

jhol opened the pull request:
This patch adds support for the OpenRISC 1000 (or1k) architecture: a MIPS-like open hardware ISA which was first introduced in 2000.

This patch adds basic architecture support for the architecture, as well as a board and soc support for the Qemu virtual ork1ksim board.

jhol commented:
@stffrdhrn Please feel free to comment. This work was the origin of the TTMR fix for Qemu.

jhol commented:
The tests will not be able to complete because the tool-chain does not have the `or1k-elf` compiler or Qemu 9.2.0.

nashif commented:
@jhol  please split into multiple commits, see https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#multiple-commits-on-a-single-pr for reference.

jhol committed:
arch: Added initial OpenRISC architecture port

This patch adds support for the OpenRISC 1000 (or1k) architecture: a
MIPS-like open hardware ISA which was first introduced in 2000.

The thread switching implementation uses the modern Zephyr thread "switch"
architecture.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
drivers: timer: Add timer driver for or1k

The OpenRISC 1000 Tick Timer is tightly coupled to the or1k CPU core, and
is explicitly designed to facilitate task scheduling and high-resolution
timing.

The timer is documented in Chapter 14 of the OpenRISC 1000 Architecture
Manual:

    https://openrisc.io/or1k.html#__RefHeading__504849_595890882

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: common: Define PERFOPT_ALIGN for OpenRISC

PERFOPT_ALIGN is defined as ".balign 4" because .balign is available in the
or1k GNU assembler.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: gcc: Added symbol attributes for OpenRISC

A common set of macros: GTEXT(), GDATA(), WTEXT() and WDATA() is defined
for Arm, Arm66, MIPS, Nios II, RISC-V and Xtensa. This patch adds OpenRISC
to this set.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: gcc: Add GEN_ABSOLUTE_SYM definitions for OpenRISC

Existing versions of GEN_ABSOLUTE_SYM and GEN_ABSOLUTE_SYM_KCONFIG are
defined for all supported architectures. This patch adds a definition of
the same kind used for MIPS, Nios II, RISC-V and Xtensa.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
openrisc: linker: Use zephyr's convention for rodata

Zephyr maps the start and end of the rodata section with variables using
the __rodata_region_start and __rodata_region_end. This patch adopts this
convention for the OpenRISC architecture.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: gcc: Define output format for or1k-elf

The OpenRISC or1k-elf GCC compiler output format is named "elf32-or1k".
This patch adds an OUTPUT_FORMAT linker macro for this platform.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
soc: qemu: Added qemu_or1k simulated SoC

Qemu includes support for the OpenRISC 1000 CPU architecture. This patch
adds a Zephyr SoC definition which enables usage of this feature.

The SoC definition closely mirrors the Qemu MIPS Malta SoC definition.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
cmake: emu: qemu: Add binary suffix for or1k architecture

The Qemu emulator executable for the OpenRISC 1000 (or1k) architecture is
named qemu-or1k. This patch adds this information to the qemu architecture
suffix list.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
boards: qemu: add board definition for the or1k architecture

Qemu includes support for the OpenRISC 1000 CPU architecture. This patch
adds a Zephyr virtual board definition which enables usage of this feature.

The board definition closely mirrors the Qemu MIPS Malta board definition.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
debug: thread_info: Store thread stack pointer offset for OpenRISC

In OpenRISC 1000, the stack pointer is stored in the r1 register. This
patch stores the offset of this value within in the thread structure into
the thread info offsets.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
tests: Add basic test-set for OpenRISC architecture port

The OpenRISC 1000 architecture can be tested through the qemu_or1k emulated
SoC and board. Because this emulated device has minimal external hardware
by default, this patch enables a minimal suite of tests covering core
kernel features for the CPU architecture.

When running the test suite, OpenRISC was found to require additional stack
space to prevent an overflow. Therefore, a minimal additional amount of
storage: 128-bytes, was added that was found in practice to allow tests to
complete.

Note that Qemu version 9.2.0 or newer is required so as to include the
following commit:

    commit 3eb43aeb164f1f83c97ff693c7d464b49755110c
    Author: Joel Holdsworth <jholdsworth@nvidia.com>
    Date:   Fri Jun 7 15:29:33 2024 -0700

        hw/openrisc: Fixed undercounting of TTCR in continuous mode

This fixes a bug in the implementation of the emulated OpenRISC Tick Timer
which prevents the Zephyr OpenRISC port from task-scheduling properly.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
scripts: logging: Add openrisc to ARCHS

The OpenRISC CPU architecture is signified by the CONFIG_OPENRISC Kconfig
option. This patch adds this information to the log database ARCHS table.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
scripts: twister: Add OpenRISC as a supported test platform

Adds the openrisc CPU architecture to the twister platform definitions.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol commented:
> please split into multiple commits

This has now been done!

jhol commented:
v3 of the patch set makes the following changes:
  * Fixed file copyrights
  * Bumped copyright year from 2024 to 2025

jhol commented:
v4 of the patch set makes the following changes:

  * Fixed warnings from `check_compliance.py`:
      * Added my name to `MAINTAINERS.yml`
      * Fixed directory paths in `MAINTAINERS.yml`
      * Replaced obsolete config defines:
          * `CONFIG_ARCH_HAS_CPU_IDLE_CUSTOM` -> `CONFIG_ARCH_HAS_CUSTOM_CPU_IDLE`
          * `CONFIG_ARCH_HAS_CPU_ATOMIC_IDLE_CUSTOM` -> `CONFIG_ARCH_HAS_CUSTOM_CPU_ATOMIC_IDLE`
      * Added missing braces
      * Fixed typos
      * Fixed a line length issue
      * Added require blank lines

jhol commented:
v4 of the patch set makes the following changes:

  * Fixed warnings from `check_compliance.py`:
      * Fixed additional blank line

jhol commented:
v6 of the patch set makes the following changes:

  * Improves the commit message of the final commit

jhol commented:
> Given this is a new architecture, can you please talk to this and give some details, background, how is this going to be tested and integrated etc in the architecture WG at some point? This will be very helpful understanding the specifics of the new architecture.

Sure, I'm happy to do so.

npitre reviewed (commented):
Some notes:

- For the purpose of `CONFIG_INIT_STACKS` you should use
  `__z_interrupt_stack_SIZEOF` as `CONFIG_ISR_STACK_SIZE` does not always cover
  the whole stack area due to rounding/alignment. Amd if SMP is ever a
  possibility then it should be `__z_interrupt_all_stacks_SIZEOF`.

- I notice your timer driver bears the signature of an earlyer timer driver
  "style". You should consider moving to the latest "style" to avoid some
  issues with the previous one. Take riscv_machine_timer.c as a good example to
  follow. Also, if you can dispense with a variable `cyc_per_tick` then do it.
  A compile time constant let the compiler optimize those costly divisions.

jhol commented:
> @jhol would you be able to join the Architecture WG next week to talk about this? cc @henrikbrixandersen

Sure! - I'd be happy to do so. Do I need to do anything to get on the agenda?

jhol commented:
@stffrdhrn Here is the info about the meeting: https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group

It seems the issue is already listed on the To-Do board here, so I think that is sufficient:
https://github.com/orgs/zephyrproject-rtos/projects/35/views/1

jhol commented:
A PR for tool-chain has been submitted here: https://github.com/zephyrproject-rtos/sdk-ng/pull/864

jhol committed:
manifest: Update picolibc to 1.8.9-2

Previously, Zephyr used a forked version of picolibc which added a patch to
disable LTO support (82d62ed1). Since an improved version of this patch has
been accepted into the the main branch in e09e6935, the forked version of
picolibc is no longer required.

Additionally, the 1.8.9 release series adds support for the OpenRISC (or1k)
architecture.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol commented:
Here is v8 of the patch-set with the following changes:

  - Rebased on `main`
  - Upgrade picolibc to the new release which officially supports OpenRISC, and no longer requires a patch via a Zephyr patch.
  - Replaced used of `CONFIG_ISR_STACK_SIZE` with `__z_interrupt_all_stacks_SIZEOF` in stack initialization
  - Fixed a typo in a comment: `s/OPENRISC/OpenRISC/`
  - Fixed instances of incorrect usage of delay slots in assembly code
  - Improved comments in assembly code
  - Fixed typos in file header comments
  - Set frame pointer before entering `z_prep_c`
  - Replaced the terms "vector" and "ISR" with "exception" and "exception handler" which are more appropriate for the architecture.
  - Renamed `_Fault` function to `z_openrisc_fault`. This change was applied to multiple other architectures in https://github.com/zephyrproject-rtos/zephyr/pull/84535
  - Moved interrupt handling loop into C
  - Added `chain_call` and `chain_call_reg` assembly macros to simplify interrupt and syscall handling code
  - Added `__reset` and `__start` symbols to `_reset_and__start` and start section as aliases of the reset exception handler
  - Added support for `CONFIG_SOC_RESET_HOOK`
  - Keep D and I-Cache enabled in new threads
  - Set `FLASH_SIZE` and `FLASH_BASE_ADDRESS` on openrisc
  - Fixed non-build of `irq_offload.c`
  - Fixed `#ifndef _ASMLANGUAGE` in `arch.h`
  - Improved linker script with more features including support for XIP

jhol commented:
@npitre I applied your suggestion concerning `__z_interrupt_stack_SIZEOF`.

Concerning the timer driver, it's not so clear to me what needs to happen here. To me, the OpenRISC timer driver doesn't look so very different from the Risc-V driver, but I'm happy to make any amendments.

npitre commented:
> Concerning the timer driver, it's not so clear to me what needs to happen
> here. To me, the OpenRISC timer driver doesn't look so very different from
> the Risc-V driver, but I'm happy to make any amendments.

You may look at commit 0ea64b3ecb64 for explanation and the needed changes,
mainly to `sys_clock_set_timeout()`.

Also, if your `sys_clock_hw_cycles_per_sec()` resolves to a constant, then
you should consider not making your `cyc_per_tick` into a variable otherwise
everywhere you use it as a divisor it will induce important runtime overhead.
If it is a constant then the compiler will most likely optimize the division
at compile time with much simpler ALU instructions.

nashif commented:
@jhol I managed to build the toolchain and qemu and was trying this PR. was able to build, but did not get any output from qemu:


`west build -p -b qemu_or1k samples/synchronization/  -t run  -- -DCONFIG_MINIMAL_LIBC=y`

picolibc did not work for me, not sure what is going on:

```
/home/nashif/bin/zephyr-sdk-0.17.0/or1k-zephyr-elf/bin/../lib/gcc/or1k-zephyr-elf/12.2.0/../../../../or1k-zephyr-elf/bin/ld.bfd: /home/nashif/bin/zephyr-sdk-0.17.0/or1k-zephyr-elf/bin/../lib/gcc/../../picolibc/or1k-zephyr-elf/lib/./libc.a: error adding symbols: bad value
collect2: error: ld returned 1 exit status
``
@keith-packard  ^^


```
Memory region         Used Size  Region Size  %age Used
             RAM:       32056 B       128 MB      0.02%
        IDT_LIST:          0 GB         2 KB      0.00%
Generating files from /home/nashif/zephyrproject/zephyr/build/zephyr/zephyr.elf for board: qemu_or1k
[120/121] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: or1k

```

jhol commented:
Thanks for testing. What version of qemu do you have?


```
Note that Qemu version 9.2.0 or newer is required so as to include the following commit:

    commit 3eb43aeb164f1f83c97ff693c7d464b49755110c
    Author: Joel Holdsworth <jholdsworth@nvidia.com>
    Date: Fri Jun 7 15:29:33 2024 -0700
    hw/openrisc: Fixed undercounting of TTCR in continuous mode

This fixes a bug in the implementation of the emulated OpenRISC Tick Timer which prevents the Zephyr OpenRISC port from task-scheduling properly.
```

nashif commented:
> Thanks for testing. What version of qemu do you have?

9.2.1, you can also get it from here https://github.com/zephyrproject-rtos/sdk-ng/actions/runs/13472947957/artifacts/2635045345 (PR https://github.com/zephyrproject-rtos/sdk-ng/pull/871)

jhol commented:
> > Thanks for testing. What version of qemu do you have?
> 
> 9.2.1, you can also get it from here https://github.com/zephyrproject-rtos/sdk-ng/actions/runs/13472947957/artifacts/2635045345 (PR https://github.com/zephyrproject-rtos/sdk-ng/pull/871)

Awesome. I'll test it out when I have time in the coming days

stffrdhrn commented on arch/openrisc/core/reboot.c:
```diff
@@ -0,0 +1,29 @@
+{
+	ARG_UNUSED(type);
+
+	__asm__("l.nop 13");
```
Note in QEMU the openrisc `virt` platform does provide reboot via the `syscon-reboot` device.  I guess we could add support for `virt` later?

carlescufi commented:
Hi @jhol . Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know of your availability here and I will schedule a date for you.

henrikbrixandersen commented:
Architecture WG 2025-02-18:
- @jhol gave a good introduction to the proposed port of Zephyr to the OpenRISC architecture.
- NVIDIA is interested in backing this architecture port, and expect to be able to back it for at least 5 years.
- The Zephyr SDK need to be updated to support the OpenRISC architecture (compiler suite, emu, …).

zgliu799 commented:
@jhol Thanks for your zephyr openrisc port. It is working well in my mor1kx and marocchino  FPGA board with just minor update.
Not sure why the  zephyr openrisc port is not in merged in the master.
CC @stffrdhrn

jhol commented:
**This PR has now been resurrected in #98160**
